src/Analyzer.PowerShellRuleEngine/PowerShellRuleEngine.cs

```diff
@@ -0,0 +1,118 @@
+
+            powerShell.Commands.AddCommand("Set-ExecutionPolicy")
+                .AddParameter("Scope", "Process") // Affects only the current PowerShell session
+                .AddParameter("ExecutionPolicy", "Unrestricted");
+            powerShell.AddStatement();
+
+            powerShell.Commands.AddCommand("Import-Module")
+                .AddParameter("Name", @".\TTK\arm-ttk.psd1"); // arm-ttk is added to the needed project's bins directories in build time 
+            powerShell.AddStatement();
+
+            powerShell.Commands.AddCommand("Test-AzTemplate")
+                .AddParameter("Test", "deploymentTemplate")
+                .AddParameter("TemplatePath", templateFilePath);
+
+            var executionResults = powerShell.Invoke();
+
+            var evaluations = new List<PowerShellRuleEvaluation>();
+
+            foreach (dynamic executionResult in executionResults)
+            {
+                var uniqueErrors = new Dictionary<string, SortedSet<int>>(); // Maps error messages to a sorted set of line numbers
+
+                foreach (dynamic warning in executionResult.Warnings)
+                {
+                    AddErrorToDictionary(warning, ref uniqueErrors);
+                }
+
+                foreach (dynamic error in executionResult.Errors)
+                {
+                    AddErrorToDictionary(error, ref uniqueErrors);
+                }
+
+                foreach (KeyValuePair<string, SortedSet<int>> uniqueError in uniqueErrors)
+                {
+                    var evaluationResults = new List<PowerShellRuleResult>();
+                    foreach (int lineNumber in uniqueError.Value)
+                    {
+                        var evaluationResult = new PowerShellRuleResult(false, lineNumber);
+                        evaluationResults.Add(evaluationResult);
+                    }
+                    var evaluation = new PowerShellRuleEvaluation(executionResult.Name, uniqueError.Key, false, evaluationResults);
+                    evaluations.Add(evaluation);
+                }
+            }
+
+            return evaluations;
+        }
+
+        private static void AddErrorToDictionary(dynamic error, ref Dictionary<string, SortedSet<int>> uniqueErrors)
+        {
+            var lineNumber = 0;
+
+            Type errorType = error.GetType();
+            IEnumerable<PropertyInfo> errorProperties = errorType.GetRuntimeProperties();
+            if (errorProperties.Where(prop => prop.Name == "TargetObject").Any())
+            {
+                if (error.TargetObject is PSObject targetObject && targetObject.Properties["lineNumber"] != null)
+                {
+                    lineNumber = error.TargetObject.lineNumber;
+                }
+            }
+
+            var lineNumberRegex = new Regex(@"\son\sline:\s\d+");
+            var errorMessage = lineNumberRegex.Replace(error.ToString(), string.Empty); 
+
+            if (uniqueErrors.ContainsKey(errorMessage))
```

Comment:
You can use TryAdd here. It simply returns false instead of throwing an error if there is a conflict in the dictionary.

Reply:
I'm trying to add something to the value of the key, not a new value. If I understand correctly from [here](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.dictionary-2.tryadd?view=net-5.0) TryAdd is used to avoid collisions when you want to do one add per key